Add helpers to construct the fallback for rich replies

Frinksy opened the pull request:
Add helpers to construct the fallback for rich replies,
both unformatted and formatted versions.

Resolves: #37

Apologies for the delay as I've been busy recently. This is what I've come up with until now, and I have two queries to move forwards:
Firstly, I'd like to know if this is actually going in the right direction to solve the issue.
Secondly, I'm not sure how to deal with `_Custom` events, the spec says to treat them the same as `m.text`, but I'm unsure about how I'd get the event's body.

Frinksy committed:
Add helpers to construct the fallback for rich replies

Add helpers to construct the fallback for rich replies,
both unformatted and formatted versions.

jplatte commented:
This is looking good but I don't think this is the way this should be exposed to users. Could you add two additional constructors based on `text_html` and `notice_html` (`text_reply` and `notice_reply`) with an extra `&MessageEventContent` or maybe `&MessageType` parameter which also sets `relates_to`?

Re. `body` I guess it might be sensible to eagerly deserialize that since every message event is supposed to have it, but for now it could just be `custom_ev_content.data["body"].as_str().unwrap_or("")`.

Frinksy committed:
Add constructors to construct replies to messages.

Add text_reply_plain, text_reply_html, notice_reply_plain,
notice_reply_html.

Frinksy committed:
Add nightly variant in rich reply fallback match statement

Also fix formatting of html

Frinksy commented:
I added a total of 4 constructors: `text_reply_plain`, `text_reply_html` and equivalents for notices.

I didn't find a way to get an `EventId` from `&MessageEventContent` or `&MessageType` to set `relates_to`, so I stuck with `&MessageEvent`.
I've also followed you suggestion for custom events and added replies to `VerificationRequest` for `unstable-pre-spec`

There are now a couple more commits, but I imagine we can squash them before merging?

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+        }
+    }
+
+    fn get_plain_quote_fallback(original_message: &MessageEvent) -> String {
```
Is there a particular reason you moved these helper functions into `impl MessageEventContent` (making them associated functions rather than free ones)?

Frinksy replied:
There is no real reason, apart from the fact I thought it "made sense" to put them under the same namespace. The last commit makes them free functions again (so that I can easily revert it if it isn't needed for them to be free).
Since they are private functions, would it still be beneficial to add a comment explaining their purpose?

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+                format!("> <{:?}> sent a video.", original_message.sender)
+            }
+            MessageType::_Custom(content) => {
+                // Eagerly deserialize `body` since we assume every message should have it
```
There is no deserialization going on here, just accessing an entry in a map.

Frinksy replied:
Yes, I had misread an earlier comment. Removed the comment.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+            }
+            #[cfg(feature = "unstable-pre-spec")]
+            MessageType::VerificationRequest(content) => {
+                // Eagerly deserialize `body` since we assume every message should have it
```
Accidental copy-paste?

Frinksy replied:
Removed the comment.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+        sent an audio file.
+    </blockquote>
+</mx-reply>",
+                    original_message.room_id.server_name(),
```
I'm pretty sure this need to be the entire room ID, not just the server name.

Frinksy replied:
You're right, changed to be the room ID.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+                format!(
+                    "<mx-reply>
+    <blockquote>
+        <a href=\"https://matrix.to/#/!{}/${}\">In reply to</a>
```
The `!` and `$` will get duplicated when interpolating the room ID & event ID like this, they shouldn't be part of the format string.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+                    original_message.room_id.server_name(),
+                    original_message.event_id.as_str(),
+                    original_message.sender.as_str(),
+                    original_message.sender.as_str(),
```
The `.as_str()` calls shouldn't be needed.

Frinksy replied:
Removed the unneeded `.as_str()` calls.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+                    original_message.event_id.as_str(),
+                    original_message.sender.as_str(),
+                    original_message.sender.as_str(),
+                )
```
You can the same `format!` argument multiple times by explicitly using the index, i.e.

```rust
        <a href=\"https://matrix.to/#/!{0}/${1}\">In reply to</a>
        <a href=\"https://matrix.to/#/{2}\">{2}</a>
```

or by simply naming one or all of the parameters:

```suggestion
                format!(
                    "<mx-reply>
    <blockquote>
        <a href=\"https://matrix.to/#/{room_id}/{event_id}\">In reply to</a>
        <a href=\"https://matrix.to/#/{sender}\">{sender}</a>
        <br />
        sent an audio file.
    </blockquote>
</mx-reply>",
                    room_id = original_message.room_id,
                    event_id = original_message.event_id,
                    sender = original_message.sender,
                )
```

You may also notice I removed the extra `!` and `$` from the format string, which would have been duplicated, as well as removing

jplatte replied:
Also could you look into replacing the `format!` with `formatdoc!` ([from indoc](https://docs.rs/indoc/1.0.3/indoc/macro.formatdoc.html)) so the indentation doesn't look so weird?

Frinksy replied:
Replaced all the `format!` calls with`formatdoc!` ones.
Also changed every call to use named arguments for readability.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +85,337 @@ impl MessageEventContent {
+            }),
+            #[cfg(feature = "unstable-pre-spec")]
+            new_content: None,
+        }
```
This could be simplified, right?

```suggestion
        Self {
            relates_to: Some(Relation::Reply {
                in_reply_to: InReplyTo { event_id: original_message.event_id.clone() },
            }),
            ..Self::text_plain(body)
        }
```

and likewise for the other new constructors

Frinksy replied:
Thank you for the reminder of this. I simplified the 4 constructors.

Frinksy committed:
Change get_html_quote_fallback to use formatdoc!

Use formatdoc! instead of format! for prettier indentation.
Address comments to simplify formatdoc! arguments.

Frinksy committed:
Simplify reply constructors.

Frinksy committed:
Make private reply fallback helpers free functions.

jplatte commented on ruma-events/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ ruma-serde = { version = "0.3.1", path = "../ruma-serde" }
 serde = { version = "1.0.118", features = ["derive"] }
 serde_json = { version = "1.0.60", features = ["raw_value"] }
 pulldown-cmark = { version = "0.8", default-features = false, optional = true }
+indoc = "1.0"
```
Please keep the dependencies ordered alphabetically.

Frinksy replied:
Moved dependency.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -11,6 +11,8 @@ use ruma_serde::StringEnum;
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
 use serde_json::Value as JsonValue;
 
+use indoc::formatdoc;
```
Please move up into the second import group.

Frinksy replied:
Moved import up.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -85,6 +87,73 @@ impl MessageEventContent {
         Self::new(MessageType::Notice(NoticeMessageEventContent::html(body, html_body)))
     }
+
+    /// A constructor to create a plain text reply to a message.
```
```suggestion
    /// Creates a plain text reply to a message.
```

jplatte replied:
Same below.

Frinksy replied:
Changed all 4 documentation strings.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -690,3 +759,235 @@ pub struct CustomEventContent {
+                room_id = original_message.room_id,
+                event_id = original_message.event_id,
+                sender = original_message.sender,
+            }
```
I know the indoc docs use curly braces but I'd much prefer this to use parentheses like usually with `format!` too.

```suggestion
            formatdoc!(
                "
                <mx-reply>
                    <blockquote>
                        <a href=\"https://matrix.to/#/{room_id}/{event_id}\">In reply to</a>
                        <a href=\"https://matrix.to/#/{sender}\">{sender}</a>
                        <br />
                        sent an audio file.
                    </blockquote>
                </mx-reply>
                ",
                room_id = original_message.room_id,
                event_id = original_message.event_id,
                sender = original_message.sender,
            )
```

jplatte replied:
Same below.

Frinksy replied:
Changed to parentheses.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -690,3 +759,235 @@ pub struct CustomEventContent {
+                    formatted.body.as_str()
+                } else {
+                    content.body.as_str()
+                }
```
This same `if let` is used in a few places here, would be good to factor it out into a function, something like

```suggestion
                body = formatted_or_plain_body(&content.formatted, &content.body),
```

Frinksy replied:
Applied the suggestion.

jplatte commented on ruma-events/src/room/message.rs:
```diff
@@ -690,3 +759,235 @@ pub struct CustomEventContent {
+                room_id = original_message.room_id,
+                event_id = original_message.event_id,
+                sender = original_message.sender,
+                body = content.data["body"].as_str().unwrap_or("")
```
Please consistently add trailing commas to format / formatdoc arguments.

```suggestion
                body = content.data["body"].as_str().unwrap_or(""),
```

Frinksy replied:
Added commas where they should be.

jplatte reviewed (commented):
Looks good overall, there's just some code style things I'd like to see improved.

Also needs a rebase since we recently moved all the crate directories into a `crates` subdirectory. Let me know if there's any issues with that.

Frinksy committed:
Address styling issues.

jplatte commented on crates/ruma-events/src/room/message.rs:
```diff
@@ -690,3 +758,242 @@ pub struct CustomEventContent {
+    if let Some(formatted_body) = formatted {
+        &formatted_body.body
+    } else {
+        &body
```
As per clippy message:

```suggestion
        body
```

jplatte replied:
This should be more obvious through a failing CI job, something seems broken there (see #562).

jplatte commented on crates/ruma-events/src/room/message.rs:
```diff
@@ -690,3 +758,242 @@ pub struct CustomEventContent {
+                        <br />
+                        {body}
+                    </blockquote>
+                </mx-reply>",
```
Last style nit:

```suggestion
                </mx-reply>
                ",
```

here and below.

jplatte commented:
Thanks!
